src/ini.zig

```diff
@@ -0,0 +1,1076 @@
+                        1 => brk: {
+                            break :brk &[_]u8{c};
+                        },
+                        2 => brk: {
+                            defer i += 1;
+                            break :brk &[_]u8{ c, val[i + 1] };
+                        },
+                        3 => brk: {
+                            defer i += 2;
+                            break :brk &[_]u8{ c, val[i + 1], val[i + 2] };
+                        },
+                        4 => brk: {
+                            defer i += 3;
+                            break :brk &[_]u8{ c, val[i + 1], val[i + 2], val[i + 3] };
+                        },
+                        // this means invalid utf8
+                        else => unreachable,
+                    }),
+                }
+            }
+
+            if (esc)
+                try unesc.append('\\');
+
+            switch (usage) {
+                .section => {
+                    this.commitRopePart(ropealloc, &unesc, &rope);
+                    return rope.?;
+                },
+                .value => {
+                    if (!did_any_escape) return Expr.init(E.String, E.String.init(val[0..]), Loc{ .start = offset });
+                    if (unesc.items.len <= STACK_BUF_SIZE) return Expr.init(
+                        E.String,
+                        E.String.init(try this.arena.allocator().dupe(u8, unesc.items[0..])),
+                        Loc{ .start = offset },
+                    );
+                    return Expr.init(E.String, E.String.init(unesc.items[0..]), Loc{ .start = offset });
+                },
+                .key => {
+                    const thestr: []const u8 = thestr: {
+                        if (!did_any_escape) break :thestr try this.arena.allocator().dupe(u8, val[0..]);
+                        if (unesc.items.len <= STACK_BUF_SIZE) break :thestr try this.arena.allocator().dupe(u8, unesc.items[0..]);
+                        break :thestr unesc.items[0..];
+                    };
+                    return thestr;
+                },
+            }
+        }
+        if (comptime usage == .value) return Expr.init(E.String, E.String.init(val[0..]), Loc{ .start = offset });
+        if (comptime usage == .key) return val[0..];
+        return strToRope(ropealloc, val[0..]);
+    }
+
+    /// Returns index to skip or null if not an env substitution
+    /// Invariants:
+    /// - `i` must be an index into `val` that points to a '$' char
+    fn parseEnvSubstitution(this: *Parser, val: []const u8, start: usize, i: usize, unesc: *std.ArrayList(u8)) ?usize {
+        if (i + 1 < val.len and val[i + 1] == '{') {
+            if (i + 2 >= val.len) return null;
+
+            var found_closing = false;
+            var j = i + 2;
+            while (j < val.len) : (j += 1) {
+                switch (val[j]) {
+                    '$' => return this.parseEnvSubstitution(val, start, j, unesc),
```

Comment:
Can you add a comment with the line of code from npm's ini parser where they do nested environment subsutitions like this?